### PR TITLE
Providing fallback CSS for inlinetoolbar

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [ ] Unit tests
- [X] Manual tests

## What changes did you make?

- Adding css for `inlinetoolbar` and `balloonpanel` (inline toolbar require it to proper work). CSS base on moono-lisa skin.

I thought to force loading `moono-lisa` as fallback, but it might be confusing if user will remove those css folders thinking that are unnecessary for his configuration. That's why I add `-default` suffix and separate files.

### CODE REVIEW

File: tests/plugins/inlinetoolbar/manual/fallbackstyles.html [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	<p>Hello world</p>
+	<img src="%BASE_PATH%_assets/lena.jpg">
+	<img src="%BASE_PATH%_assets/lena.jpg">
+	<img src="%BASE_PATH%_assets/lena.jpg">
+	<img src="%BASE_PATH%_assets/lena.jpg">
+	<p>&nbsp;</p>
+	<p>&nbsp;</p>
+	<p>&nbsp;</p>
+</div>
+
```
Comment (f1ames) [inline tests/plugins/inlinetoolbar/manual/fallbackstyles.html]: You could also add inline editor here.

**Comment by msamsel** [comment]
@mlewand what do you think about that?
> This task is about `inlinetoolbar` and from what I recall the default CSS was supposed to be added only for `inlinetoolbar` plugin. If there are some additional styles from `balloonpanel` needed they probably should be placed in `inlinetoolbar` CSS file too.

Inline toolbar inheritance styles from balloon panel related to:
- drawing triangle below/above balloon
- ballon position (like setting proper display mode, z-index, etc.)

**Comment by msamsel** [comment]
I haven't notice that after rebase there was change in API so `create` method became `attach`.
I've made al changes suggested in your review. I extract from `balloonpanel` only important styles (from perspective of `inlinetoolbar`).

I have to prepare 2 separate test suites with css. Unfortunately skins cannot be changed after loading, so I have to made proper thing at the very begging of each test suite.

File: tests/plugins/inlinetoolbar/cssskin.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+				skin: 'moono'
+			}
+		}
+	};
+
+	// We need to set up skin at the very beginning befor files start to load.
+	CKEDITOR.skinName = 'moono';
+
+	var tests = {
+		'test check default css usage': function( editor ) {
```
Comment (f1ames) [inline tests/plugins/inlinetoolbar/cssskin.js]: Should it be `test check specific|moono skin css usage`?

File: tests/plugins/inlinetoolbar/manual/fallbackstyles.html [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		} );
+	}
+
+	var appendStyleSheet = sinon.stub( CKEDITOR.dom.document.prototype, 'appendStyleSheet', function( cssFileUrl ) {
+		// Simualate missing css in skin.
+		if ( cssFileUrl.indexOf( 'inlinetoolbar.css' ) > -1 || cssFileUrl.indexOf( 'balloonpanel.css' ) > -1 ) {
+			cssFileUrl = '';
+		}
+
+		if ( this.$.createStyleSheet ) {
```
Comment (f1ames) [inline tests/plugins/inlinetoolbar/manual/fallbackstyles.html]: We didn't you use call to original function here same as in unit test?

File: tests/plugins/inlinetoolbar/manual/fallbackstyles.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+@bender-ui: collapsed
+@bender-tags: 4.8.0, bug, inlinetoolbar
+@bender-ckeditor-plugins: wysiwygarea,toolbar,basicstyles,floatingspace,inlinetoolbar,link,image,resize,language,stylescombo
+
+Perform below steps in both editors:
+1. Click into image.
```
Comment (f1ames) [inline tests/plugins/inlinetoolbar/manual/fallbackstyles.md]: 1. _Click **on** image_ or just _Click image_.

File: tests/plugins/inlinetoolbar/manual/fallbackstyles.html (hunk 3 above)
Comment (msamsel) [reply tests/plugins/inlinetoolbar/manual/fallbackstyles.html]: I've made it earlier with stub and forget about it. I think option with calling origin function is simpler, and can be resistant for possible method changes.
I've update this to looks like in Unit test.

**Comment by msamsel** [comment]
As I talk with @mlewand we currently should focus on most popular browsers, and older IE left for further improvements. That's why I ignore tests on IE below 11.
I have add different assertion for the IE envs.

I also a little tune our tools method, which converts rgb to hex to also support rgba. If you consider it as separate issue, give me an info I'll extract it, but this might extend realisation of this issue. As this introduce dependancy here. Another option would be creating specific function just for those test purposes, but I'm not think is a good option, because this will duplicate the code.

File: plugins/inlinetoolbar/skins/default.css [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+/* align: [ hcenter ] */
+.cke_balloon.cke_inlinetoolbar .cke_balloon_triangle_outer.cke_balloon_triangle_align_hcenter
+{
+	left: 50%;
+	margin-left: -20px;
+}
+
+.cke_balloon.cke_inlinetoolbar
```
Comment (f1ames) [inline plugins/inlinetoolbar/skins/default.css]: This is the same selector as in line 6 (https://github.com/ckeditor/ckeditor-dev/pull/1136/files#diff-a222e1835c2e1c3b856fdaf98d45e2faR6), it should be merged into one rule.

File: plugins/inlinetoolbar/skins/default.css [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	top: -10px;
+}
+
+.cke_balloon.cke_inlinetoolbar .cke_balloon_triangle_inner.cke_balloon_triangle_top
+{
+	border-width: 0 10px 10px;
+}
+
+/* align: [ hcenter ] */
+.cke_balloon.cke_inlinetoolbar .cke_balloon_triangle_outer.cke_balloon_triangle_align_hcenter
```
Comment (f1ames) [inline plugins/inlinetoolbar/skins/default.css]: This is the same selector as in line 28 (https://github.com/ckeditor/ckeditor-dev/pull/1136/files#diff-a222e1835c2e1c3b856fdaf98d45e2faR28), it should be merged into one rule.

File: tests/plugins/inlinetoolbar/_helpers/default.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/* exported replaceAppendStyleSheet, convertRgbaToRgb */
+
+'use strict';
+
+function replaceAppendStyleSheet() {
```
Comment (f1ames) [inline tests/plugins/inlinetoolbar/_helpers/default.js]: This looks good, however it doesn't provide a possibility to clean up after running tests (restore original function). You could do two things:

* This function might return a function which will restore original one when called:
```
return function() {
    CKEDITOR.dom.document.prototype.appendStyleSheet = oldFn;
}
```
* Use sinon stub which has `stub.restore()` method.

It is good practise to always clean up after test run (especially for unit tests, for manual you may skip it). So after changing it, please add clean up to unit test (you may use `tearDown` function).

Comment (msamsel) [reply tests/plugins/inlinetoolbar/_helpers/default.js]: I decided to not use `sinon.stub`, because I wasn't able to find possibility to fire origin function within stub.
In documentation I found only possibility how to replace the method, with own function.
That's why I wrote some similar wrapper which could execute native method. I'll extend it a little bit to restore native method.

**Comment by msamsel** [comment]
I can use neither `setUp` nor `tearDown`. Those methods are executed too late. Skins are loaded when global `CKEDITOR` object is created and are applied to all editors at given page. That's why test cases are separated between 2 files, and that's why I cannot use it in setUp. Because those methods are executed after main lib is fully loaded.
That's why I made cleaning up in a different approach inside an event, when editor should have all css loaded, so modification shouldn't be required any longer.

File: tests/plugins/inlinetoolbar/_helpers/default.js (hunk 7 above)
Comment (f1ames) [reply tests/plugins/inlinetoolbar/_helpers/default.js]: It works fine but it's a little over complicated for me. You have to create new instance, it has two methods which you have to call, etc. plus you reinvented a wheel (well, a part of it:D).

> In documentation I found only possibility how to replace the method, with own function.

That's exactly where you can call the original function, the stub could look like this:

```
function stubAppendStyleSheet() {
	var originFn = CKEDITOR.dom.document.prototype.appendStyleSheet;
	return sinon.stub( CKEDITOR.dom.document.prototype, 'appendStyleSheet', function( cssFileUrl ) {
		// Simulate missing css in skin.
		if ( !cssFileUrl.match( /(inlinetoolbar|balloonpanel)\.css/ ) ) {
			originFn.call( CKEDITOR.document, cssFileUrl );
		}
	} );
}
```

and then you can use it like:
```
var stub = stubAppendStyleSheet();
CKEDITOR.once( 'instanceReady', stub.restore );
```

It is shorter and simpler IMHO. Plus for everybody who knows `sinon`, you just take a look at the usage and it should be clear that these two lines stubs `appendStyleSheet` method and then restores it. With custom code you always have to dive in to check what's going on.
